chapter10/drjt/ex5.pl

```diff
@@ -0,0 +1,53 @@
+#!/usr/bin/perl
+use 5.010;
+use warnings;
+use strict;
+
+use File::Basename;
+use File::Spec::Functions;
+use Data::Dumper;
+
+use File::Basename;
+use File::Spec::Functions;
+
+my $data = data_for_path( '../..', 'breadth-first' );
+print Dumper $data;
+
+sub data_for_path {
+  my( $path, $mode ) = @_;
+
+  my $data = {};
+  my $code;
+
+    $code = ( $mode == 'breadth-first' ) ? sub {
+      my ( $queue, @paths ) = @_;
+      push @$queue, @paths;
+    } : sub {
+      my ( $queue, @paths ) = @_;
+      push @$queue, @paths;
+    };
```

Comment:
Your queue is defined just below this - you could reduce the complexity by capturing the lexical `@queue` instead of receiving it in the arguments passed. (Also, that `$mode == 'breadth-first'` should be `$mode eq 'breadth-first'`).